File: 1950.java
import java.util.ArrayDeque;
import java.util.Deque;

public class Solution {
    public int[] findMaximums(int[] nums) {
        int n = nums.length;
        int[] left = new int[n], right = new int[n];
        Deque<Integer> stk = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            while (!stk.isEmpty() && nums[stk.peek()] >= nums[i]) {
                int top = stk.pop();
                left[top] = stk.isEmpty() ? -1 : stk.peek();
                right[top] = i;
            }
            
            stk.push(i);
        }
        
        while (!stk.isEmpty()) {
            int top = stk.pop();
            left[top] = stk.isEmpty() ? -1 : stk.peek();
            right[top] = n;
        }
        
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            int len = right[i] - left[i] - 2;
            res[len] = Math.max(res[len], nums[i]);
        }
        for (int i = n - 2; i >= 0; i--) {
            res[i] = Math.max(res[i], res[i + 1]);
        }
        
        return res;
    }
}

File: 1950.py
def findMaximums(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = []
        
        def append_to_deque(q, index):
            while q and nums[q[-1]] > nums[index]:
                q.pop()
            q.append(index)
        
        # for windows from 1 to n
        for window_size in range(1, n+1):
            start = 0 
            max_val = float('-inf')
            q = deque()
            for end in range(n):
                # monotonic stack keep track of min element in window
                append_to_deque(q, end)
                if  end-start+1 >= window_size:
                    max_val = max(max_val, nums[q[0]])
                    
                    # if element to be removed is on stack remove
                    if start == q[0]:
                        q.popleft()
                    start+=1
            res.append(max_val)            
        return res